astroid/nodes/node_classes.py

```diff
@@ -234,6 +235,13 @@ def _container_getitem(instance, elts, index, context=None):
             return new_cls
         if isinstance(index, Const):
             return elts[index.value]
+    except ValueError as exc:
+        raise AstroidValueError(
+            message="Slice {index!r} cannot index container",
+            node=instance,
+            index=index,
+            context=context,
+        ) from exc
```

Comment:
I think it would make more sense to catch it even earlier in `_infer_slice`. `0` is always an invalid value for `step`. So we shouldn't infer a `slice` object if we encounter it.
https://github.com/PyCQA/astroid/blob/4acf5785c54f4ccb8f41402991f6ddc5d8b28b89/astroid/nodes/node_classes.py#L211-L223

This would also have the benefit of fixing the other call to `_infer_slice` as well.
https://github.com/PyCQA/astroid/blob/4acf5785c54f4ccb8f41402991f6ddc5d8b28b89/astroid/nodes/node_classes.py#L2022-L2023

To reproduce, run pylint for
```py
for _ in ""[::0]:
    pass
```

Reply:
Thanks, this is a cleaner solution. I'll get to it today

Reply:
Looks like I've got a pretty generous definition of "today" these days. I think I originally considered this but stopped because it's not actually an error to _create_ a `slice` object with a step of 0, just an error to use it to index the built-in sequences:

```python
>>> slice(None, 10, 0)
slice(None, 10, 0)
>>> [][slice(None, 10, 0)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: slice step cannot be zero
```

Authors can choose to handle extended indexes however they'd like, though the only example that comes to mind is `numpy.nd_grid`: https://numpy.org/doc/stable/reference/generated/numpy.mgrid.html
```python
>>> np.mgrid[0:3, 0:3]
array([[[0, 0, 0],
        [1, 1, 1],
        [2, 2, 2]],

       [[0, 1, 2],
        [0, 1, 2],
        [0, 1, 2]]])
```

or `range` supports sub-ranges through slices:
```python
>>> range(20)
range(0, 20)
>>> range(20)[::3]
range(0, 20, 3)
>>> range(20)[:5]
range(0, 5)
```

It turns out that slice indices don't even have to be integers as I've just learned, https://docs.python.org/3/reference/datamodel.html#index-67:
> ... start is the lower bound; stop is the upper bound; step is the step value; each is None if omitted. These attributes can have any type.

Pylint has tests for these cases (such as the string indices), but at least one of them is incorrect and will need to be changed: https://github.com/PyCQA/pylint/blob/main/tests/functional/i/invalid/invalid_slice_index.py#L25-L28
```python
# Valid indices
def function4():
    """integers used as indices"""
    return TESTLIST[0:0:0] # no error
```

I think the appropriate fix is probably:
1. Does the LHS have a custom `__getitem__` method? If so, we're probably in the uninferable realm
2. Try to defer the computed indexes to `slice.indices` since this is the canonical source of the raised `Type` and `ValueError`s:

```python
>>> [][:10:0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: slice step cannot be zero
>>> slice(None, 10, 0).indices(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: slice step cannot be zero
>>> (1, 2, 3)["a":"z"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: slice indices must be integers or None or have an __index__ method
>>> slice('a', 'z').indices(26)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: slice indices must be integers or None or have an __index__ method
```

I'm leaning towards adding the `except ValueError` case in `Const.getitem` similarly what was added here in `_container_getitem`. We can refactor the duplication out (now or later) since the distinction isn't really between "const" and "container", it's between "indexables" and "sequences" https://docs.python.org/3/reference/datamodel.html#index-15.

Reply:
> I think I originally considered this but stopped because it's not actually an error to _create_ a `slice` object with a step of 0, just an error to use it to index the built-in sequences:

Good point!

> Pylint has tests for these cases (such as the string indices), but at least one of them is incorrect and will need to be changed: https://github.com/PyCQA/pylint/blob/main/tests/functional/i/invalid/invalid_slice_index.py#L25-L28
> 
> ```python
> # Valid indices
> def function4():
>     """integers used as indices"""
>     return TESTLIST[0:0:0] # no error
> ```
> 
> I think the appropriate fix is probably:
> 
> 1. Does the LHS have a custom `__getitem__` method? If so, we're probably in the uninferable realm
> 2. Try to defer the computed indexes to `slice.indices` since this is the canonical source of the raised `Type` and `ValueError`s:

Pylint does most of that already. It only emits `invalid-slice-index` for a set of known objects which expect an int-like type. We would just need to special-case the `0` step error. https://github.com/PyCQA/pylint/blob/v2.15.5/pylint/checkers/typecheck.py#L1769-L1778

> I'm leaning towards adding the `except ValueError` case in `Const.getitem` similarly what was added here in `_container_getitem`.

Sounds good.